### PR TITLE
ace_v1x: host system time set and get

### PR DESCRIPTION
**Pull request by piotrkmx**
Base firmware functionality for setting and getting system time. This is for passing the information about host system time. It is then used by FW to translate event timestamps e.g. logs to host system time domain

Signed-off-by: Piotr Kmiecik <piotrx.kmiecik@intel.com>

### CODE REVIEW

File: src/audio/base_fw.c [hunk 1]
```diff
@@ -8,6 +8,10 @@
 #include <ipc4/pipeline.h>
 #include <ipc4/logging.h>
 #include <sof_versions.h>
+#if CONFIG_ACE_V1X_ART_COUNTER
```
Comment (abonislawski) [inline src/audio/base_fw.c:11]: even without CONFIG_ACE_V1X_RTC_COUNTE?

File: src/audio/base_fw.c [hunk 2]
```diff
@@ -228,12 +234,99 @@ static int basefw_mem_state_info(uint32_t *data_offset, char *data)
...
+	return IPC4_SUCCESS;
+}
+
+static uint32_t basefw_get_ext_system_time(uint32_t *data_offset, char *data)
+{
+	uint64_t host_time = ((uint64_t)global_system_time_info.host_time.val_u << 32)
+				| (uint64_t)global_system_time_info.host_time.val_l;
+	uint64_t dsp_time = ((uint64_t)global_system_time_info.dsp_time.val_u << 32)
+				| (uint64_t)global_system_time_info.dsp_time.val_l;
+	(void)host_time;
```
Comment (abonislawski) [inline src/audio/base_fw.c:271]: whats the purpose of this part of the code without CONFIG_ACE_V1X_ART_COUNTER && CONFIG_ACE_V1X_RTC_COUNTER?

File: src/audio/base_fw.c [hunk 3]
```diff
@@ -228,12 +234,99 @@ static int basefw_mem_state_info(uint32_t *data_offset, char *data)
...
+	api->get_value_64(dev, &rtc);
+	ext_system_time_data.rtc_l = (uint32_t)rtc;
+	ext_system_time_data.rtc_u = (uint32_t)(rtc >> 32);
+
+	memcpy_s(ext_system_time, sizeof(ext_system_time), &ext_system_time_data,
+		 sizeof(ext_system_time));
+	*data_offset = sizeof(struct ipc4_ext_system_time);
+	return IPC4_SUCCESS;
+#endif
+	return IPC4_FAILURE;
```
Comment (abonislawski) [inline src/audio/base_fw.c:321]: wrong return, this is general (unknown) failure

File: src/include/ipc4/base_fw.h [hunk 4]
```diff
@@ -558,6 +558,21 @@ struct ipc4_system_time_info {
 	struct ipc4_system_time dsp_time;
 } __attribute__((packed, aligned(4)));
 
+struct ipc4_ext_system_time {
+	//! Lower DWORD of current system time value.
```
Comment (abonislawski) [inline src/include/ipc4/base_fw.h:562]: align comments style to the rest in this file

File: src/ipc/ipc4/handler.c [hunk 5]
```diff
@@ -712,8 +712,8 @@ static int ipc4_get_large_config_module_instance(struct ipc4_message_request *ip
 					&data_offset, data);
 
 	/* set up ipc4 error code for reply data */
-	if (ret < 0)
-		ret = IPC4_MOD_INVALID_ID;
+	if (ret != 0)
```
Comment (abonislawski) [inline src/ipc/ipc4/handler.c:715]: separate commit?

**Review by abonislawski — CHANGES_REQUESTED** [review]

File: src/audio/base_fw.c [hunk 6]
```diff
@@ -228,12 +234,99 @@ static int basefw_mem_state_info(uint32_t *data_offset, char *data)
...
+#if CONFIG_ACE_V1X_ART_COUNTER && CONFIG_ACE_V1X_RTC_COUNTER
+	struct ipc4_ext_system_time *ext_system_time = (struct ipc4_ext_system_time *)(data);
+	struct ipc4_ext_system_time ext_system_time_data = {0};
+
+	uint64_t hh = 0;
+	uint64_t wallclk = 0;
+	uint64_t rtc = 0;
+
+	const struct device *dev;
+	const struct counter_driver_api *api;
```
Comment (abonislawski) [inline src/audio/base_fw.c:283]: this is not needed

File: src/audio/base_fw.c [hunk 7]
```diff
@@ -228,12 +234,99 @@ static int basefw_mem_state_info(uint32_t *data_offset, char *data)
...
+
+	dev = device_get_binding(DT_LABEL(DT_NODELABEL(ace_art_counter)));
+
+	if (!dev) {
+		LOG_DBG("board: ACE V1X device binding failed");
+		return IPC4_FAILURE;
+	}
+
+	api = (const struct counter_driver_api *)dev->api;
+	api->get_value_64(dev, &hh);
```
Comment (abonislawski) [inline src/audio/base_fw.c:293]: Do not use directly counter_driver_api
counter_get_value_64()

File: src/audio/base_fw.c [hunk 8]
```diff
@@ -228,12 +234,99 @@ static int basefw_mem_state_info(uint32_t *data_offset, char *data)
...
+
+	dev = device_get_binding(DT_LABEL(DT_NODELABEL(ace_rtc_counter)));
+
+	if (!dev) {
+		LOG_DBG("board: ACE V1X device binding failed");
+		return IPC4_FAILURE;
+	}
+
+	api = (const struct counter_driver_api *)dev->api;
+	api->get_value_64(dev, &rtc);
```
Comment (abonislawski) [inline src/audio/base_fw.c:312]: counter_get_value_64()

**Review by abonislawski — CHANGES_REQUESTED** [review]

File: src/include/ipc4/base_fw.h (hunk 4 above)
Comment (mmaka1) [reply src/include/ipc4/base_fw.h:562]: It is a good idea, the IPC4 interface requires _Doxygen & Breathe_ compatible documentation, but it should be introduced as a separate PR for the entire file.

File: src/audio/base_fw.c [hunk 9]
```diff
@@ -8,6 +8,10 @@
 #include <ipc4/pipeline.h>
```
Comment (kv2019i) [inline src/audio/base_fw.c:8]: I'm a bit puzzled. System time is maintained by Zephyr. I don't see this PR setting the Zephyr system time, or modifying it. If Zephyr system time is not modified, who would use this information? E.g. log timestamps come from Zephyr uptime, not from SOF IPC applicatin code.

File: src/audio/base_fw.c (hunk 3 above)
Comment (kv2019i) [reply src/audio/base_fw.c:321]: And breaks integration with kernel. If we report an error, the kernel will refuse to work with the FW. So we need to silently ignore this if unimplemented.

**Review by kv2019i — CHANGES_REQUESTED** [review]
The commit requires a bit more explanation. The commit prefix is "drivers:" but there is no driver code here.

File: src/audio/base_fw.c (hunk 9 above)
Comment (piotrkmx) [reply src/audio/base_fw.c:8]: Hi, as far as I understand from ACE documentation, this is for passing the information about host system time. It is then used by FW to translate event timestamps e.g. logs to host system time domain. And you are right. Prefix is a little bit unfortunate. Just a leftover from separation into SOF and Zephyr part. I'll update the title and description.

File: src/include/ipc4/base_fw.h [hunk 10]
```diff
@@ -558,6 +558,21 @@ struct ipc4_system_time_info {
...
+	uint32_t utc_l;
+	//! Upper DWORD of current system time value.
+	uint32_t utc_u;
+	//! Lower DWORD of current RTC value.
+	uint32_t rtc_l;
+	//! Upper DWORD of current RTC value.
+	uint32_t rtc_u;
+	//! Lower DWORD of current HH value.
+	uint32_t hh_l;
```
Comment (mmaka1) [inline src/include/ipc4/base_fw.h:571]: `hh` should not be used in a common interface definition as it is Intel HW specific name. `art` (always running timer) seems to be more appropriate.

File: src/audio/base_fw.c [hunk 11]
```diff
@@ -228,12 +234,99 @@ static int basefw_mem_state_info(uint32_t *data_offset, char *data)
...
-				   uint32_t param_id,
-				   bool first_block,
-				   bool last_block,
-				   uint32_t *data_offset,
-				   char *data)
+			uint32_t param_id,
+			bool first_block,
+			bool last_block,
+			uint32_t *data_offset,
+			char *data)
```
Comment (softwarecki) [inline src/audio/base_fw.c:329]: You break code formatting

File: src/audio/base_fw.c (hunk 6 above)
Comment (piotrkmx) [reply src/audio/base_fw.c]: @abonislawski Please check if that's what you meant.

File: src/ipc/ipc4/handler.c [hunk 12]
```diff
@@ -717,8 +717,8 @@ static int ipc4_get_large_config_module_instance(struct ipc4_message_request *ip
...
```
Comment (kv2019i) [inline src/ipc/ipc4/handler.c:720]: Not blocking the PR, but why this change? We use "ret < 0" check everywhere else in the codebase, so it looks a bit odd to use a different check in on place. Just check sof/src/ipc/ipc4/handler.c

File: src/audio/base_fw.c [hunk 13]
```diff
@@ -228,13 +235,99 @@ static int basefw_mem_state_info(uint32_t *data_offset, char *data)
...
+				       bool last_block,
+				       uint32_t data_offset,
+				       char *data)
+{
+	if (!(first_block && last_block))
+		return IPC4_INVALID_REQUEST;
+
+	global_system_time_info.host_time.val_l = (*(struct ipc4_system_time *)(data)).val_l;
+	global_system_time_info.host_time.val_u = (*(struct ipc4_system_time *)(data)).val_u;
+	uint64_t current_dsp_time = sof_cycle_get_64();
```
Comment (kv2019i) [inline src/audio/base_fw.c:249]: @piotrkmx @mwasko @abonislawski Can you help me out here. I don't understand why we set global_system_time_info (and use SRAM to store this in DSP global memory) if it's only used by host to ask for host time. This makes no sense as host already knows its own time. 

Probably the intent is that we should configure Zephyr's system time to global_system_time_info (or only Zephyr logging subsystem's time), so this action would also has some impact to DSP operation. Is there other purpose than affecting logging timestamps? That will impact whether we should set Zephyr's kernel time, or is it enough to set timing offset for the logging subsystem. Based on my initial review, setting the kernel time might be more complex (I see no such interfaces, need to consult Zephyr folks. adding @nashif to the loop here ).

File: src/audio/base_fw.c [hunk 14]
```diff
@@ -228,13 +235,99 @@ static int basefw_mem_state_info(uint32_t *data_offset, char *data)
...
+		LOG_DBG("board: ACE V1X device binding failed");
+		return IPC4_FAILURE;
+	}
+
+	counter_get_value_64(dev, &art);
+
+	wallclk = sof_cycle_get_64();
+	ext_system_time_data.art_l = (uint32_t)art;
+	ext_system_time_data.art_u = (uint32_t)(art >> 32);
+	uint64_t delta = (wallclk - dsp_time) / (CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC / 1000000);
```
Comment (kv2019i) [inline src/audio/base_fw.c:295]: Now this get_ext_system_time() part of the PR I do understand. This querying a clock value the host otherwise does not have access to.

**Review by kv2019i — CHANGES_REQUESTED** [review]
The code itself looks clean, no issues with that, but I'm still missing the overall intent of the get/set system time. If this is intended to affect log timestamps, then this PR is only a partial implementation. Please see comments inline.

File: src/audio/base_fw.c [hunk 15]
```diff
@@ -228,13 +235,99 @@ static int basefw_mem_state_info(uint32_t *data_offset, char *data)
...
+	ext_system_time_data.utc_u = (uint32_t)(new_host_time >> 32);
+
+	dev = DEVICE_DT_GET(DT_NODELABEL(ace_rtc_counter));
+
+	if (!dev) {
+		LOG_DBG("board: ACE V1X device binding failed");
+		return IPC4_FAILURE;
+	}
+
+	counter_get_value_64(dev, &rtc);
```
Comment (mmaka1) [inline src/audio/base_fw.c:309]: Make sure this value is provided by RTC HAL function on Intel platform.

File: src/audio/base_fw.c [hunk 16]
```diff
@@ -228,13 +235,99 @@ static int basefw_mem_state_info(uint32_t *data_offset, char *data)
...
+	uint64_t rtc = 0;
+
+	const struct device *dev = DEVICE_DT_GET(DT_NODELABEL(ace_art_counter));
+
+	if (!dev) {
+		LOG_DBG("board: ACE V1X device binding failed");
+		return IPC4_FAILURE;
+	}
+
+	counter_get_value_64(dev, &art);
```
Comment (mmaka1) [inline src/audio/base_fw.c:290]: Make sure this value is provided by ART HAL function on Intel platforms.

File: src/audio/base_fw.c [hunk 17]
```diff
@@ -228,13 +235,99 @@ static int basefw_mem_state_info(uint32_t *data_offset, char *data)
...
+	return IPC4_SUCCESS;
+}
+
+static uint32_t basefw_get_ext_system_time(uint32_t *data_offset, char *data)
+{
+#if CONFIG_ACE_V1X_ART_COUNTER && CONFIG_ACE_V1X_RTC_COUNTER
+	struct ipc4_ext_system_time *ext_system_time = (struct ipc4_ext_system_time *)(data);
+	struct ipc4_ext_system_time ext_system_time_data = {0};
+
+	uint64_t host_time = ((uint64_t)global_system_time_info.host_time.val_u << 32)
```
Comment (mmaka1) [inline src/audio/base_fw.c:272]: This logic assumes that global_system_time_info had been set by another IPC call. Should we respond with error to report wrong IPC flow in case this value is uninitialized?

File: src/audio/base_fw.c (hunk 13 above)
Comment (kv2019i) [reply src/audio/base_fw.c:249]: Or is this only set to calculate the ART/RTC value later in ext_get_system_time()?

Comment (abonislawski) [reply src/audio/base_fw.c:249]: @kv2019i the usage is very limited in this PR but it can be used in another PR for logging timestamp offset, e.g using log_set_timestamp_func() which already exists in zephyr.
I think we dont want to change zephyr kernel time.

Comment (piotrkmx) [reply src/audio/base_fw.c:249]: @kv2019i 
Yes, before ext get system time system time can be used, it must be set via system time. 

Of course the host knows it's host time. If I understand correctly, the purpose of this is to let Zephyr know the host time. As Zephyr system time can be different from host system time. So it can use this information to produce logs that include not Zephyr's kernel time, but with host system time.  The other usage mentioned besides logs, is probes packets. 

I personally do not see the point to change the whole Zephyr's kernel time, so firmware can produce logs or any other event timestamps that contain in themselves the current host system time. As we discussed it today within our team, it may have way to much implications.

Comment (kv2019i) [reply src/audio/base_fw.c:249]: Ack thanks. I do agree that changing the kernel looks like potentially high-effort and strictly not needed. @piotrkmx can you add a "/* TODO: configurate time to logging subsystem */" which would make this clearer in the code.

File: src/audio/base_fw.c [hunk 18]
```diff
@@ -228,13 +235,99 @@ static int basefw_mem_state_info(uint32_t *data_offset, char *data)
...
+				| (uint64_t)global_system_time_info.host_time.val_l;
+	uint64_t dsp_time = ((uint64_t)global_system_time_info.dsp_time.val_u << 32)
+				| (uint64_t)global_system_time_info.dsp_time.val_l;
+	(void)host_time;
+	(void)dsp_time;
```
Comment (kv2019i) [inline src/audio/base_fw.c:277]: Btw, in this version 'host_time' and 'dsp_time' are both used in this function, so these definitions on L276 and L277 are no longer needed, right=

File: src/ipc/ipc4/handler.c (hunk 12 above)
Comment (piotrkmx) [reply src/ipc/ipc4/handler.c:720]: There was and issue with 'ret < 0'. drv->ops.get_large_config() can return positive or negative error code. For some reason, EINVAL is returned as a negative value, but e.g. IPC4_UNAVAILABLE is positive. I did not want to change error codes. 0 is always a success, so I figured out that this would be the best solution. There are other entities of 'ret != 0' in handler.c. After set_pipeline_state() and cpu_enable_core(). I do not insist on this solution, but I need a suggestion, how can we resolve this problem.

Comment (kv2019i) [reply src/ipc/ipc4/handler.c:720]: Thanks @piotrkmx . I misunderstood the change indeed. We indeed have two error numbering schemes. SOF has traditionally used the POSIX model of negative error codes. So the root problem is that get_large_config() may return either type of error. If it's a IPC4 error (positive), it can be passed directly as a return value (which expects an IPC4 type error). Only if the function returns a negative value, we need to convert it from POSIX to IPC4. So I believe the original code is correct and this change is not needed.

File: src/audio/base_fw.c (hunk 15 above)
Comment (piotrkmx) [reply src/audio/base_fw.c:318]: @mmaka1 Please correct me if I'm wrong, but I think that the whole point of what we're doing here is to abandon those HAL functions in favor of Zephyr legitimate drivers. In case of RTC and ART, two additional Intel drivers have been added to Zephyr (PR: https://github.com/zephyrproject-rtos/zephyr/pull/49175). They are supposed to do the same thing as the HAL functions that you've mentioned. To ensure that we are getting the right values, we pass the the proper device as an argument of counter_get_value_64() function.

File: src/audio/base_fw.c (hunk 16 above)
Comment (piotrkmx) [reply src/audio/base_fw.c:299]: @mmaka1 Please correct me if I'm wrong, but I think that the whole point of what we're doing here is to abandon those HAL functions in favor of Zephyr legitimate drivers. In case of RTC and ART, two additional Intel drivers have been added to Zephyr (PR: https://github.com/zephyrproject-rtos/zephyr/pull/49175). They are supposed to do the same thing as the HAL functions that you've mentioned. To ensure that we are getting the right values, we pass the the proper device as an argument of counter_get_value_64() function.

File: src/audio/base_fw.c (hunk 17 above)
Comment (piotrkmx) [reply src/audio/base_fw.c:280]: Response added. Please just check, if those kind of verification it's sufficient.

File: src/ipc/ipc4/handler.c (hunk 12 above)
Comment (lyakh) [reply src/ipc/ipc4/handler.c]: @kv2019i a slight correction: I think `.get_large_config()` is supposed to **only** return negatice POSIX errors or 0, never positive IPC4 errors. Mixing both in the same function would be bad. And no, I still don't think that even mixing both in the same subsystem is a good idea.

Comment (mwasko) [reply src/ipc/ipc4/handler.c]: > @kv2019i a slight correction: I think `.get_large_config()` is supposed to **only** return negatice POSIX errors or 0, never positive IPC4 errors. Mixing both in the same function would be bad. And no, I still don't think that even mixing both in the same subsystem is a good idea.

@lyakh IPC processing functions must implement a way to return an IPC specific status codes. I am not saying that current implementation is ideal, but we need something. The POSIX here is insufficient.

File: src/audio/base_fw.c [hunk 19]
```diff
@@ -9,13 +9,25 @@
...
 DECLARE_SOF_RT_UUID("basefw", basefw_comp_uuid, 0xe398c32, 0x5ade, 0xba4b,
 		    0x93, 0xb1, 0xc5, 0x04, 0x32, 0x28, 0x0e, 0xe4);
 DECLARE_TR_CTX(basefw_comp_tr, SOF_UUID(basefw_comp_uuid), LOG_LEVEL_INFO);
 
+static struct ipc4_system_time_info global_system_time_info = {
+		.host_time.val_l = 0,
+		.host_time.val_u = 0,
+		.host_time.val_l = 0,
+		.host_time.val_u = 0
+};
```
Comment (kv2019i) [inline src/audio/base_fw.c:29]: I believe this is unnecessary. BSS section will be zeroed.

**Review by kv2019i — APPROVED** [review]
Minor nit on initialization of global data, but nothing to prevent a merge. Looks good!

File: src/audio/base_fw.c (hunk 19 above)
Comment (piotrkmx) [reply src/audio/base_fw.c]: @kv2019i ok, I wasn't sure before. Removed.

File: src/audio/base_fw.c (hunk 15 above)
Comment (mmaka1) [reply src/audio/base_fw.c:313]: Oh, I left my both comments regarding counter_get_value_64() API incomplete, sorry for that. I meant '... while working on other platforms too'. Are we OK with the current implementation where this IPC4 call fails on platforms other than Intel ACE?

**Review by marcinszkudlinski — APPROVED** [review]

File: src/audio/base_fw.c (hunk 15 above)
Comment (kv2019i) [reply src/audio/base_fw.c:313]: @mmaka1  That's a good point. So basicly the code to look up which device to use for counter_get_value_64(), should be platform specific, and ideally behind a bit of abstraction to make this scale when more platforms are added.

Considering this is the first platform to support this, and we return IPC4_UNAVAILABLE, I think this is still ok starting point and the generalization of the interface can be done in follow-up PRs. @mwasko @lgirdwood what do you think?

**Comment by kv2019i** [comment]
Status: CI and reviews ok, one open raised by @mmaka1 w.r.t. how extended time is tied to Intel ace1x platform, is open. Please see the inline discussion above.

File: src/audio/base_fw.c [hunk 20]
```diff
@@ -228,13 +235,103 @@ static int basefw_mem_state_info(uint32_t *data_offset, char *data)
...
+		return IPC4_INVALID_RESOURCE_STATE;
+
+	uint64_t art = 0;
+	uint64_t wallclk = 0;
+	uint64_t rtc = 0;
+
+	const struct device *dev = DEVICE_DT_GET(DT_NODELABEL(ace_art_counter));
+
+	if (!dev) {
+		LOG_DBG("board: ACE V1X device binding failed");
```
Comment (abonislawski) [inline src/audio/base_fw.c:290]: This log can inform which device was not found

File: src/audio/base_fw.c [hunk 21]
```diff
@@ -228,13 +235,103 @@ static int basefw_mem_state_info(uint32_t *data_offset, char *data)
...
+		return IPC4_FAILURE;
```
Comment (abonislawski) [inline src/audio/base_fw.c:291]: Do not use IPC4_FAILURE in this file

**Review by abonislawski — CHANGES_REQUESTED** [review]

File: src/audio/base_fw.c (hunk 15 above)
Comment (abonislawski) [reply src/audio/base_fw.c:313]: This is using common api (counter_get_value_64) so already it is not platform specific, if the device didn't have ace in the name everything would be ok, just the implementation is platform specific but this is internal zephyr thing, like for dma we have platform specific implementations but one generic device with common api, not sure why there is ace everywhere but I dont know history of this driver.

File: src/audio/base_fw.c (hunk 21 above)
Comment (piotrkmx) [reply src/audio/base_fw.c]: @abonislawski Please check if it's ok now.

File: src/audio/base_fw.c (hunk 15 above)
Comment (piotrkmx) [reply src/audio/base_fw.c:313]: The history behind is that it was named ace following the naming convention of other Zephyr drivers (interrupt drivers to be more precise, which introduced the distinction between platforms). I agree that this was not the best choice, but we have discussed this within our team and the conclusion is that it does not matter that much at this point and for this PR. To avoid further confusion, those Zephyr  drivers will be renamed to intel_art_counter and intel_rtc_counter in the future.

**Review by abonislawski — APPROVED** [review]

**Review by tmleman — APPROVED** [review]

File: src/audio/base_fw.c (hunk 15 above)
Comment (kv2019i) [reply src/audio/base_fw.c:313]: @abonislawski @piotrkmx It's not just the "ace" part. Code in "src/audio/" should be vendor neutral, so this should rely on just a generic counter device that is used in the IPC4 device counter (could be non-ART on non-Intel platforms). But alas, this is the first hardware that adds the support, so I agree this can be generalized in later PRs when/if the need comes. FYI @lgirdwood

File: src/audio/base_fw.c [hunk 22]
```diff
@@ -228,13 +235,103 @@ static int basefw_mem_state_info(uint32_t *data_offset, char *data)
...
+	const struct device *dev = DEVICE_DT_GET(DT_NODELABEL(ace_art_counter));
+
+	if (!dev) {
+		LOG_DBG("board: ART counter device binding failed");
+		return IPC4_MOD_NOT_INITIALIZED;
+	}
+
+	counter_get_value_64(dev, &art);
+
+	wallclk = sof_cycle_get_64();
```
Comment (lgirdwood) [inline src/audio/base_fw.c:296]: This can be fixed incrementally, we need to use the Zephyr native API here too for wallclock (i.e. abstraction will be at the Zephyr level)

**Review by lgirdwood — APPROVED** [review]
